ros2service/test/test_echo.py

```diff
@@ -0,0 +1,167 @@
+    pytest.skip(
+            'CLI tests can block for a pathological amount of time on Windows.',
+            allow_module_level=True)
+
+
+@pytest.mark.rostest
+@launch_testing.parametrize('rmw_implementation', get_available_rmw_implementations())
+def generate_test_description(rmw_implementation):
+    path_to_introspectable_script = os.path.join(
+        os.path.dirname(__file__), 'fixtures', 'introspectable.py'
+    )
+    additional_env = {'RMW_IMPLEMENTATION': rmw_implementation}
+    return LaunchDescription([
+        # Always restart daemon to isolate tests.
+        ExecuteProcess(
+            cmd=['ros2', 'daemon', 'stop'],
+            name='daemon-stop',
+            on_exit=[
+                ExecuteProcess(
+                    cmd=['ros2', 'daemon', 'start'],
+                    name='daemon-start',
+                    on_exit=[
+                        # Add test fixture actions.
+                        Node(
+                            executable=sys.executable,
+                            arguments=[path_to_introspectable_script],
+                            name='introspectable_service',
+                            additional_env=additional_env,
+                        ),
+                        launch_testing.actions.ReadyToTest()
+                    ],
+                    additional_env=additional_env
+                )
+            ]
+        ),
+    ])
+
+
+class TestROS2ServiceEcho(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(
+        cls,
+        launch_service,
+        proc_info,
+        proc_output,
+        rmw_implementation
+    ):
+        @contextlib.contextmanager
+        def launch_service_command(self, arguments):
+            service_command_action = ExecuteProcess(
+                cmd=['ros2', 'service', *arguments],
+                additional_env={
+                    'RMW_IMPLEMENTATION': rmw_implementation,
+                    'PYTHONUNBUFFERED': '1'
+                },
+                name='ros2service-echo',
+                output='screen'
+            )
+            with launch_testing.tools.launch_process(
+                launch_service, service_command_action, proc_info, proc_output,
+                output_filter=launch_testing_ros.tools.basic_output_filter(
+                    filtered_prefixes=[
+                        'waiting for service to become available...',
+                        '/launch_ros'  # cope with launch_ros internal node.
+                    ],
+                    filtered_rmw_implementation=rmw_implementation
+                )
+            ) as service_command:
+                yield service_command
+        cls.launch_service_command = launch_service_command
+
+    @launch_testing.markers.retry_on_failure(times=5, delay=1)
+    def test_echo_no_arr(self):
+        echo_arguments = ['echo', '/test_introspectable', '--no-arr']
+        expected_output = [
+            'info:',
+            '  event_type: 0',
```

Comment:
True.  That would be nicer.  The main reason we aren't doing something like that is that we are reusing [message_to_yaml](https://github.com/ros2/rosidl_runtime_py/blob/82af486e09d5463c2c4cc5135d2e122735d4f2be/rosidl_runtime_py/convert.py#L66), and it automatically converts the fields to integers for us.

The only ways I can think of to do this nicer here would be to:
1. Make a local copy of `message_to_yaml` which understands to convert this type to a meaningful name rather than a number.
2. Take the string that `message_to_yaml` returns, and do a text substitution of `request_type: <int>` to `request_type: <meaningful_name>`.

They both seem pretty ugly to me, but maybe you have another idea on how to improve this?

Reply:
Alternatively, call [message_to_ordereddict](https://github.com/ros2/rosidl_runtime_py/blob/82af486e09d5463c2c4cc5135d2e122735d4f2be/rosidl_runtime_py/convert.py#L162) make the necessary change to that structure and print with `yaml.dump`.

Reply:
The above solution mentioned by @jacobperron seems good to me.

There is no relationship between `_event_type` and `Constants` in the `build/service_msgs/rosidl_generator_py/service_msgs/msg/_service_event_info.py`, it seems impossible to only update the `rosidl_generator_py/resource/_idl.py.em` to achieve this goal.
(NOTE: Maybe supporting `Enum` in the `rosidl` is the final solution.)

Since it needs to update another [repository](https://github.com/ros2/rosidl_runtime_py.git), I think a follow-up improvement for https://github.com/ros2/ros2cli/pull/745#discussion_r1195719282 might be better.

Reply:
> Alternatively, call [message_to_ordereddict](https://github.com/ros2/rosidl_runtime_py/blob/82af486e09d5463c2c4cc5135d2e122735d4f2be/rosidl_runtime_py/convert.py#L162) make the necessary change to that structure and print with yaml.dump.

Good idea.  I did a prototype of that locally, and that seems to work.

However, now I'm questioning whether we should do this.  One of the benefits of the current representation (with the numbers) is that it is valid `ServiceEventInfo` message, so it is possible to save the output from `ros2 service echo` to a file, load it into a YAML parser (like `yaml.safe_load`), and then have it be something you could make a new message with.  If we change the `event_type` representation to be a string, that will no longer be the case.

But I honestly don't know if that is a valid use case.  What do other people think?

Reply:
IMO,

> it is possible to save the output from ros2 service echo to a file, load it into a YAML parser (like yaml.safe_load), and then have it be something you could make a new message with.

this should be done via `rosbag2`, not `ros2 service echo` to peek the service activity. after all, i am with @jacobperron suggestion on this one.

Reply:
At the risk of adding too much complexity, maybe there could be an (opt-in or opt-out) option to make the output more human friendly. Not only could it translate the event type integers to names, but it could pretty up some other output, like the client uuid (it could print a hash or something less verbose to the console).

Reply:
> At the risk of adding too much complexity, maybe there could be an (opt-in or opt-out) option to make the output more human friendly. Not only could it translate the event type integers to names, but it could pretty up some other output, like the client uuid (it could print a hash or something less verbose to the console).

After thinking about it more, I agree with @fujitatomoya that the ability to copy the data from stdin and have it be valid YAML isn't that interesting; there are other tools to do that.  So with that said, I've pushed my prototype implementation that does this replacement and makes it more human readable.  I don't think we need the extra complexity for having extra modes here.

In terms of making the GID less verbose, one thing that we *could* do here is make flow-style the default.  That means that instead of:

```
  client_gid:
  - 1
  - 15
  - 112
  ...
```

we'll print out:

```
client_gid: [1, 15, 112, 44, 3, 108, 235, 199, 0, 0, 0, 0, 0, 0, 18, 4]
```

Which isn't necessarily more readable, but is far more compact.  The only downside to doing that is that we then differ in behavior from `ros2 topic echo`, which does *not* use flow-style by default.

Overall, I'm inclined to just make the change to the `event_type` and leave everything else as-is.  But I'm happy to hear other opinions.

Reply:
> Overall, I'm inclined to just make the change to the event_type and leave everything else as-is. But I'm happy to hear other opinions.

SGTM